### DISCUSSION TITLE
Why do we have to "Specify the Web SDK in the project file" when doing integration tests in ASP.NET Core?

### DISCUSSION THREAD

**Discussion by edumserrano** [dotnet/aspnetcore#49601] [Q&A] [answered]
On the [Test app prerequisites](https://learn.microsoft.com/en-us/aspnet/core/test/integration-tests?view=aspnetcore-7.0#test-app-prerequisites) section of the `Integration tests in ASP.NET Core` docs it says:

> The test project must:
>
> - Reference the [Microsoft.AspNetCore.Mvc.Testing](https://www.nuget.org/packages/Microsoft.AspNetCore.Mvc.Testing) package.
> - Specify the Web SDK in the project file (\<Project Sdk="Microsoft.NET.Sdk.Web">).

I couldn't find anywhere information about why we need to "specify Web SDK in the project file" ? I'm just curious about this as I can also get integration tests to run whilst leaving the SDK set to `Sdk="Microsoft.NET.Sdk"`.

**Comment by martincostello** [comment, marked as answer]
The reference means that some MSBuild targets will be pulled in that will copy files needed by things such as MVC and Razor Pages if you're doing UI tests (I think there's some other use cases too, but I forget what they are right now).

I would imagine that it was easier for the document to be written with _"add this one line"_ rather than _"add this one line if you do any of these X use cases_".

If you remove it and your tests all still pass then I don't think there's any harm in you leaving it out, but for the sake of one line and potential future confusion it's probably best just to add it so things should "Just Work ™️" as your application and tests evolve over time.

**Comment by edumserrano** [comment]
Thanks, I was wondering what those edge cases would be where it is absolutely required. 

I do agree that:
> it's probably best just to add it so things should "Just Work ™️" as your application and tests evolve over time.

However, the side effect is dealing with things like a `launchsettings.json` getting created when you change the SDK. Found this article that helped me out [How to stop Visual Studio from creating launchsettings.json](https://andrewlock.net/how-to-stop-visual-studio-from-creating-launchsettings-json/). To be fair this is the only side effect I've encountered so far.

It would probably help if that part of the documentation was clearer on when you need to change the SDK in the project file. And I also think that adding a note about the `NoDefaultLaunchSettingsFile` property would be positive.

@martincostello, thank you for your reply. If possible, I'd like to get better clarification on :
> I would imagine that it was easier for the document to be written with "add this one line" rather than "add this one line if you do any of these X use cases".

So, I'm going to leave this question open for now, hoping someone can better detail what are the pros/cons to using `Microsoft.NET.Sdk` vs `Microsoft.NET.Sdk.Web` for integration tests.

**Reply by martincostello** [reply]
Maybe @Rick-Anderson has an opinion on this as it's mainly a documentation question (I think)?

**Comment by dnperfors** [comment]
I think it is mostly because of the default references and global usings that come with the sdk. Technically it is not necessary and personally I avoid it.
